feat: cookbook v1

realeinherjar opened the pull request:
- [x] Convert `build.sh` to a GitHub Actions and closes rust-bitcoin/www.rust-bitcoin.org#6.
- [x] Automated CI in GitHub Actions that test all the code in the cookbook: `.github/workflows/test.yml`
- [x] Automated scheduled CI in GitHub Actions that check all markdown files for broken links (we are adding a LOT of links to the `rust-bitcoin` docs): `.github/workflows/check-links.yml`
- [x] Dependabot to update (weekly) `.github/dependabot.yml`:
  - [x] GitHub Actions
  - [x] Test dependencies, this is good once we release `rust-bitcoin` new versions dependabot will open a PR here to update
- [x] `justfile` to easy run the build and test with `just`
- [x] ~~This is how you receive data over P2P~~ moved to rust-bitcoin/rust-bitcoin.github.io#9.
- [x] ~~This is how you parse blocks and transactions~~ moved to rust-bitcoin/rust-bitcoin.github.io#8.
- [x] This is how you construct and sign transactions
  - [x] [TABConf 2023 segwit signing exercise](https://github.com/tcharding/workshop/tree/master/sign-segwit-v0)
  - [x] [TABConf 2023 taproot signing exercise](https://github.com/tcharding/workshop/tree/master/sign-taproot)

## Details

I want this cookbook to be tested automatically on CI, so the code will always run.
The only solution that worked for me was [this workaround](https://github.com/rust-lang/mdBook/issues/706#issuecomment-1139423009) that is used in [The Rust Rand Book](https://github.com/rust-random/book).

realeinherjar committed:
feat: cookbook

Cookbook on how to sign segwit and taproot transactions

Additionally:

- Convert `build.sh` to a GitHub Action
- Automated CI in GitHub Actions that test all the code in the cookbook
- Automated scheduled CI in GitHub Actions that check all markdown files for broken links
- Dependabot to update (weekly)
  - GitHub Actions
  - Test dependencies
- `justfile` to easy run the build and test with `just`

realeinherjar commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,13 @@
+    schedule:
+      # Check for updates to GitHub Actions every week
+      interval: "weekly"
+  - package-ecosystem: "cargo"
+    directory: "cookbook/tests/"
+    schedule:
+      # Check for updates to Rust dependencies every week
+      interval: "weekly"
```
NOTE to myself: I need to test this (first time using the `package-ecosystem: "cargo"`).

realeinherjar replied:
Done in a99f4ebcd58d55e01ccfef07eddd099355bf8985

realeinherjar commented on .github/workflows/gh-pages.yml:
```diff
@@ -0,0 +1,25 @@
+
+      - run: mdbook build
+
+      - name: Deploy
+        uses: peaceiris/actions-gh-pages@v3
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          publish_dir: ./cookbook/book
```
@tcharding this is totally wrong, I am adding as a placeholder now.
I'll update once I finish to add the content.
Feel free to comment, edit, contribute...

tcharding replied:
I'm very far from a GitHub actions expert, please take my lack of comments on the actions stuff as "I don't know enough to comment right now but I don't see anything too crazy"

realeinherjar replied:
I will update this soon with the correct code and any questions or if you feel something is odd let me know...

realeinherjar replied:
@tcharding the CI action for gh-pages deployment is ready to review. Should work now.

realeinherjar replied:
Double-checked once again and made slight tweaks. This should work now.

realeinherjar commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,25 @@
+name: Tests
+
+on:
+  push:
+    branches: [ master, '0.[0-9]+' ]
+  pull_request:
+    branches: [ master, '0.[0-9]+' ]
```
NOTE: we might use `v0.[0-9]+`

realeinherjar replied:
done in 3c531c6482451bba7ee8d075ab004bbab572c1cc

realeinherjar commented on cookbook/src/03-parse.md:
```diff
@@ -0,0 +1 @@
+# Parsing Blocks and Transactions
```
@tcharding any code snippet up your sleeve that I can deconstruct and explain in simple steps?

tcharding replied:
Not off the top of my head, if I were doing this I'd probably start by grepping the code base for `Transaction::from` looking for hex being parsed then go from there. Same for other types. If you want to add demos of code hitting bitcoin-core you could look at https://github.com/Kixunil/pico-bitcoin-wallet (or my version it `github.com/tcharding/workshop/pico-bitcoin-wallet`).

tcharding replied:
Be warned, the `FromHex` trait has moved out of `hashes` and into `hex-conservative` as of the soon-to-be-released code. If you want to code against `rust-bitcoin v0.30.0` I can do the upgrade easily after we release `v0.31.0` or you can code against `master` and we change it after release to v0.31.0 - what ever is easiest for you.

realeinherjar replied:
What if we follow `rust-bitcoin#master` and I run the tests CI once a week?
That would make the cookbook most up-to-date with the codebase

tcharding replied:
I think we want to get to a stage where the cookbook is running released code, most users will be using released code. I only mentioned master because we are about to release (soon TM)

realeinherjar commented on cookbook/src/02-p2p.md:
```diff
@@ -0,0 +1 @@
+# Receiving Data over P2P
```
@tcharding any code snippet up your sleeve that I can deconstruct and explain in simple steps?

tcharding replied:
None, I have not played around with p2p stuff very much at all.

realeinherjar replied:
Ok I will remove it for now from the PR, we can revisit later

tcharding commented on cookbook/book.toml:
```diff
@@ -1,7 +1,7 @@
 authors = [
     "Harshil Jani <harshiljani2002@gmail.com>",
-    "Tobin Harding<me@tobin.cc>",
+    "Tobin C. Harding<me@tobin.cc>",
```
Thanks!

tcharding commented:
Ready for review, right? Perhaps squash it all down into a single commit, its going to be a punish to rebase otherwise. Since its the first major work on the cookbook its fine to be one enormous commit (at least I'm fine to review it as such). I got super excited when I ran `mdbook serve` and saw what you'd done, I've wanted to make progress on this for so long, thanks a million man!

(FTR I'm going to review it now as is still.)

apoelstra commented:
+1 to squashing into one commit -- though if you wanna be really nice, you could maybe split it into sections and make each one its own commit, or something.

tcharding commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,28 @@
+name: Tests
+on:
+  push:
+    branches: [ master, '0.[0-9]+', 'v0.[0-9]+' ]
```
These are regex's, right? Isn't the third one redundant since anything that matches the second one will match the third? (FTR I don't know if branch `this-is-not-master` matches the first one).

realeinherjar replied:
`this-is-not-master` will not match, because we are not using regex in `master`.
This makes `master` an exact match.

For the second one, yes you are probably right.
Additionally I think we should change that to match any version apart from `0.x`.

Done in 04bf29508c5c98be51217e3ecbce58a1fa6994fd

tcharding commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,28 @@
+          toolchain: stable
+      - name: Generate harness
+        working-directory: ./cookbook/tests
+        run: ./generate.sh
```
This is sick!

tcharding commented on README.md:
```diff
@@ -1,4 +1,6 @@
-# Website source for www.rust-bitcoin.org
+# Website source for <www.rust-bitcoin.org>
+
+[![CC0 1.0][cc-shield]][cc]
```
Attention to detail, I love it.

tcharding commented on cookbook/book.toml:
```diff
@@ -1,6 +1,17 @@
 [book]
-authors = ["Harshil Jani"]
+authors = [
+    "Harshil Jani <harshiljani2002@gmail.com>",
```
@Harshil-Jani are you ok to have your email address added here?

Harshil-Jani replied:
Yupp.

tcharding commented on cookbook/src/intro.md:
```diff
@@ -0,0 +1,47 @@
+[`rust-bitcoin`](https://github.com/rust-bitcoin/rust-bitcoin) is a library for working with Bitcoin in Rust.
+It contains Bitcoin network protocol and associated primitives.
+You can find more by reading the [documentation](https://docs.rs/bitcoin)
+and by visiting [rust-bitcoin.org](https://rust-bitcoin.org).
```
This reads a little odd to me because this cookbook is hosted at rust-bitcoin.org

tcharding commented on cookbook/src/intro.md:
````diff
@@ -0,0 +1,47 @@
+Here's an example:
+
+```bash
+cargo add bitcoin --features=no-std
````
"no-std" is not a good example to use here (because we break Rust convention, this feature cannot be use without turning off default features, bad rust-bitcoin devs, no biscuit). 

Perhaps use "rand-std" as an example feature?

realeinherjar replied:
Good catch, done in fad21c36599224836e48cd0aebaf81128a3bf60d

tcharding commented on cookbook/src/intro.md:
```diff
@@ -0,0 +1,47 @@
+This book contains:
+
+1. [Constructing and Signing Transactions](tx.md)
+    1. [SegWit V0](tx_segwit-v0.md)
+    1. [Taproot](tx_taproot.md)
```
TIL, I didn't know using 1 repeated worked - nice!

realeinherjar replied:
Yes, it is a markdown good practice, make the diffs way nicer when you are adding stuff to a list.

tcharding commented on cookbook/src/tx.md:
```diff
@@ -0,0 +1,6 @@
+We provide the following examples:
+
+- [Constructing and Signing Transactions - SegWit V0](04a-tx_segwit-v0.md)
+- [Constructing and Signing Transactions - Taproot](04b-tx_taproot.md)
```
The links are wrong here. The ones in the intro work
```
    1. [SegWit V0](tx_segwit-v0.md)
    1. [Taproot](tx_taproot.md)
```

realeinherjar replied:
Good catch.
Done in 9995dadaa6385fd61f70f7e8a54025eee292e9b9

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+# Constructing and Signing Transactions - SegWit V0
+
+In this section, we will construct a [SegWit V0 transaction](https://bitcoinops.org/en/topics/segregated-witness/).
```
Awesome choice of reference.

tcharding commented on cookbook/src/tx_segwit-v0.md:
````diff
@@ -0,0 +1,342 @@
+
+Next, we define the following constants:
+
+```rust
````
It would be nice if we could get syntax highlighting, using `mdbook serve` I don't get it. https://rust-lang.github.io/mdBook/format/theme/syntax-highlighting.html seems to say that we should get it by default since we use '```rust'?

realeinherjar replied:
This is strange... What version of `mdbook` are you using?
This is what I get from a `mdbook serve` in the `tx_segwit-v0` page:
<img width="791" alt="image" src="https://github.com/rust-bitcoin/www.rust-bitcoin.org/assets/136860407/8221e5e1-d30b-4df4-af97-cff13a9345ec">

tcharding replied:
Oh cool, not to worry then. FWIW `mdbook v0.4.25`

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+Hence, it is necessary to import the `std::str::FromStr` trait.
+Note that `bc1q7cyrfmck2ffu2ud3rn5l5a8yv6f0chkp0zpemf` is a [Bech32](https://bitcoinops.org/en/topics/bech32/) address.
+This is an arbitrary, however valid, Bitcoin mainnet address,
+since we use the `require_network` method to ensure that the address is valid for the mainnet.
```
Perhaps?
```suggestion
This is an arbitrary, however valid, Bitcoin mainnet address.
Hence we use the `require_network` method to ensure that the address is valid for mainnet.
```

tcharding commented on cookbook/src/tx_segwit-v0.md:
````diff
@@ -0,0 +1,342 @@
+```
+
+`dummy_unspent_transaction_output` generates a dummy unspent transaction output (UTXO).
+This is a SegWit V0 P2WPKH (`ScriptBuf::new_v0_p2wpkh`) UTXO with a dummy invalid transaction ID (`txid: Txid::all_zeros()`),
````
hmmm, this is going to go stale in next release. I mention it because it highlights that this cookbook is going to take some careful maintenance to make  sure it doesn't go stale. Perhaps we should hack up a script that pulls out everything between code ticks then we can look it over each release and see if anything jumps out at us?

realeinherjar replied:
We will be testing this extensively.

We are now only testing for `bitcoin@v0.30`:

https://github.com/rust-bitcoin/www.rust-bitcoin.org/blob/b4bc52864a9144333933b800536c197dddf70fbf/cookbook/tests/Cargo.toml#L9

However, we have a dependabot routine that will update weekly the dependencies in `Cargo.toml`, which includes `bitcoin`:

https://github.com/rust-bitcoin/www.rust-bitcoin.org/blob/b4bc52864a9144333933b800536c197dddf70fbf/.github/dependabot.yml#L9-L13

So this would open a dependabot PR for updating `bitcoin` to a new version and then we would test the cookbook code against a new version of `bitcoin`.
This is possible because of:

https://github.com/rust-bitcoin/www.rust-bitcoin.org/blob/b4bc52864a9144333933b800536c197dddf70fbf/.github/workflows/test.yml#L14-L28

tcharding replied:
Yeah but the function name I commented on just appears in text, its not in a rust code block. By maintenance burden I meant that, as it stands, we really have to re-read the whole cookbook each release and make sure none of the mentioned functions have changed names.

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+We are using the [`OutPoint`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.OutPoint.html) struct to represent the transaction output.
+Finally, we return the tuple `(out_point, utxo)`.
+
+Now we are ready for our main function that will sign a transaction that spends an `p2wpkh` unspent output:
```
```suggestion
Now we are ready for our main function that will sign a transaction that spends a `p2wpkh` unspent output:
```

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+- `sequence` is the sequence number; it is a [`Sequence`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.Sequence.html) type.
+   We are using the [`ENABLE_RBF_NO_LOCKTIME`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.Sequence.html#associatedconstant.ENABLE_RBF_NO_LOCKTIME) constant.
+- `witness` is the witness stack; it is a [`Witness`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.Witness.html) type.
+   We are using the [`default`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.Witness.html#impl-Default) method.
+   This is possible because `Witness` implements the [`Default`](https://doc.rust-lang.org/std/default/trait.Default.html) trait.
```
Perhaps it would be better to say something like "we use default to create an empty witness that will be filled in later after signing"

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+- `value` is the amount we are spending; it is a [`u64`](https://doc.rust-lang.org/std/primitive.u64.html) type.
+   We are using the `const SPEND_AMOUNT` that we defined earlier.
+- `script_pubkey` is the script code required to spend a P2WPKH output; it is a [`ScriptBuf`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/script/struct.ScriptBuf.html) type.
+   We are using the [`script_pubkey`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/address/struct.Address.html#method.script_pubkey) method to generate the script code.
```
```suggestion
   We are using the [`script_pubkey`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/address/struct.Address.html#method.script_pubkey) method to generate the script pubkey from the receivers address.
```

(Not sure just this second what form we are using for "script pubkey" (ie., "scriptPubkey").

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+- `input` is the input vector; it is a [`Vec<TxIn>`](https://doc.rust-lang.org/std/vec/struct.Vec.html) type.
+   We are using the `input` variable that we defined earlier wrapped in the [`vec!`](https://doc.rust-lang.org/std/macro.vec.html) macro for convenient initialization.
+- `output` is the output vector; it is a [`Vec<TxOut>`](https://doc.rust-lang.org/std/vec/struct.Vec.html) type.
+   We are using the `spend` and `change` variables that we defined earlier wrapped in the [`vec!`](https://doc.rust-lang.org/std/macro.vec.html) macro for convenient initialization.
```
Perhaps mention that order does not matter but by convention it goes `[spend, change]` (I think that is the convention anyways).

realeinherjar replied:
I didn't understand what you meant

tcharding replied:
In the code the outputs vector is `[spend, change]` but could have been `[change, spend]`. Both are correct (I believe), when I wrote this I spent a little time looking up if there was a convention on this and it seems that `[spend, change]` is it but I'm not 100% sure (I could be mixing it up with the fact that in derivation path 0 is external and 1 is change) .

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+   We are using the `spend` and `change` variables that we defined earlier wrapped in the [`vec!`](https://doc.rust-lang.org/std/macro.vec.html) macro for convenient initialization.
+
+In `let mut sighash_cache = SighashCache::new(unsigned_tx);` we are instantiating a [`SighashCache`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html) struct.
+This is a type that ffficiently calculates [signature hash message](https://developer.bitcoin.org/devguide/transactions.html?highlight=sighash_all#signature-hash-types) for legacy, segwit and taproot inputs.
```
typo "efficiently"

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+We are using the `new` method to instantiate the struct with the `unsigned_tx` that we defined earlier.
+`new` takes any `Borrow<Transaction>` as an argument.
+[`Borrow<T>`](https://doc.rust-lang.org/std/borrow/trait.Borrow.html) is a trait that allows us to pass either a reference to a `T` or a `T` itself.
+Hence, you can pass a `Transaction` or a `&Transaction` to `new`.
```
Interesting, I would have expected this to mention `BorrowMut` but for some reason the code works without a mutable borrow - I don't actually understand how we are calling `tx.borrow_mut()` when the tx is only `Borrow` - bizzare. No action required, just mentioning.

realeinherjar replied:
I think that
```rust
let mut sighash_cache = SighashCache::new(unsigned_tx);
```

actually takes ownership of `unsigned_tx`

We are not calling borrow under the hood,
see: https://docs.rs/bitcoin/latest/src/bitcoin/crypto/sighash.rs.html#557-559

tcharding replied:
> actually takes ownership of `unsigned_tx`

Correct but the tx is of type `R: Borrow<Transaction>`

> We are not calling borrow under the hood,

On line 1080 of the same page you link to we call `self.tx.borrow_mut()`

https://docs.rs/bitcoin/latest/src/bitcoin/crypto/sighash.rs.html#1079

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+[`Borrow<T>`](https://doc.rust-lang.org/std/borrow/trait.Borrow.html) is a trait that allows us to pass either a reference to a `T` or a `T` itself.
+Hence, you can pass a `Transaction` or a `&Transaction` to `new`.
+
+`sighash_cache` is instantiated as mutable because we will be updating it by adding a [`segwit_signature_hash`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.segwit_signature_hash) to it.
```
This is not exactly true, perhaps
```
`sighash_cache` is instantiated as mutable because we require a mutable reference when creating the sighash to sign using [`segwit_signature_hash`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.segwit_signature_hash).

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+   We are using the `const DUMMY_UTXO_AMOUNT` that we defined earlier.
+- `sighash_type` is the type of sighash; it is a [`EcdsaSighashType`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html) enum.
+   We are using the [`All`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html#variant.All) variant,
+   that sign all the outputs.
```
```suggestion
   which indicates that the sighash will include all the outputs.
```

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+   We are using the [`All`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html#variant.All) variant,
+   that sign all the outputs.
+
+We compute the message `msg` by converting the `sighash` to a [`Message`](https://docs.rs/secp256k1/latest/secp256k1/struct.Message.html) type.
```
```suggestion
We create the message `msg` by converting the `sighash` to a [`Message`](https://docs.rs/secp256k1/latest/secp256k1/struct.Message.html) type.
```

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+   that sign all the outputs.
+
+We compute the message `msg` by converting the `sighash` to a [`Message`](https://docs.rs/secp256k1/latest/secp256k1/struct.Message.html) type.
+This is a hashed message input to an ECDSA signature.
```
```suggestion
 This is the message that we will sign.
```

realeinherjar replied:
Can we remove the empty space ` ` in the beginning of the line?

tcharding replied:
Seems so, typo by me I'd say.

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+
+We compute the message `msg` by converting the `sighash` to a [`Message`](https://docs.rs/secp256k1/latest/secp256k1/struct.Message.html) type.
+This is a hashed message input to an ECDSA signature.
+The [Message::from](https://docs.rs/secp256k1/latest/secp256k1/struct.Message.html#impl-From%3C%26%27_%20bitcoin%3A%3Ahashes%3A%3Asha256d%3A%3AHash%3E) method takes anything that implements the [`Hash` trait](https://docs.rs/bitcoin_hashes/latest/bitcoin_hashes/trait.Hash.html) as an argument[^ThirtyTwoByteHash].
```
```suggestion
The [Message::from](https://docs.rs/secp256k1/latest/secp256k1/struct.Message.html#impl-From%3C%26%27_%20bitcoin%3A%3Ahashes%3A%3Asha256d%3A%3AHash%3E) method takes anything that promises to be a thirty two byte hash i.e., 32 bytes that came from a cryptographically secuer hashing algorithm.
```

realeinherjar replied:
There's a typo... `secuer` -> `secure`

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+It takes a refence to a [`Message`](https://docs.rs/secp256k1/latest/secp256k1/struct.Message.html) and a reference to a [`SecretKey`](https://docs.rs/secp256k1/latest/secp256k1/struct.SecretKey.html) as arguments,
+and returns a [`Signature`](https://docs.rs/secp256k1/latest/secp256k1/ecdsa/struct.Signature.html) type.
+
+In the next step, we update the `sighash_cache` first with [`witness_mut`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.witness_mut) method.
```
Perhaps?
```suggestion
In the next step, we update the witness stack for the input we just signed using the `SighashCache::witness_mut`(https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.witness_mut) method.
```

realeinherjar replied:
This would not render a markdown link, we are missing the brackets `[ ]`.

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+and returns a [`Signature`](https://docs.rs/secp256k1/latest/secp256k1/ecdsa/struct.Signature.html) type.
+
+In the next step, we update the `sighash_cache` first with [`witness_mut`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.witness_mut) method.
+This allows us to modify the witness stack of the input we are signing.
```
And drop this one.

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+In the next step, we update the `sighash_cache` first with [`witness_mut`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.witness_mut) method.
+This allows us to modify the witness stack of the input we are signing.
+It returns an [`Option<&mut Witness>`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/witness/struct.Witness.html) type.
+Next, the witness stack is by using the [`push_bitcoin_signature`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/witness/struct.Witness.html#method.push_bitcoin_signature) method.
```
```suggestion
We use the [`push_bitcoin_signature`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/witness/struct.Witness.html#method.push_bitcoin_signature) method.
```

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+   which returns a `SerializedSignature` type.
+1. A [`EcdsaSighashType`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html) enum.
+   Again we are using the same [`All`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html#variant.All) variant that we used earlier.
+
```
Here we need to push the public key as well (this was a bug in my solution to the workshop found by @sanket1729, an attempt to improve the API is made here: https://github.com/rust-bitcoin/rust-bitcoin/pull/2084

realeinherjar replied:
Ok I am following the PR. From my understanding, this does not change the code we will show in the cookbook.

tcharding replied:
The code is missing the push of the pubkey, I commented on the line of code where its needed. The PR I linked to adds a function so that one does not have to push the sig and keys separately but it won't be there till next release.

realeinherjar replied:
fixed

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+1. A [`EcdsaSighashType`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html) enum.
+   Again we are using the same [`All`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html#variant.All) variant that we used earlier.
+
+Finally, we convert the `sighash_cache` back to a [`Transaction`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.Transaction.html) type by using the [`into_transaction`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.into_transaction) method.
```
```suggestion
Finally, we destroy the `sighash_cache` and get out the signed [`Transaction`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.Transaction.html) type by using the [`into_transaction`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.into_transaction) method.
```

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+Finally, we convert the `sighash_cache` back to a [`Transaction`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.Transaction.html) type by using the [`into_transaction`](https://docs.rs/bitcoin/latest/bitcoin/sighash/struct.SighashCache.html#method.into_transaction) method.
+
+As the last step we print this to terminal using the [`println!`](https://doc.rust-lang.org/std/macro.println.html) macro.
+This transaction is now ready to be broadcasted to the Bitcoin network.
```
```suggestion
This transaction is now ready to be broadcast to the Bitcoin network.
```

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,342 @@
+
+[^expect]: We will be unwraping any [`Option<T>`](https://doc.rust-lang.org/std/option)/[`Result<T, E>`](https://doc.rust-lang.org/std/result) with the `expect` method.
+
+[^ThirtyTwoByteHash]: Additionally it is also necessary that the hash is 32-bytes long.
```
If you decide to use my suggestion then this goes away. I'm not sure we should bother folk at this level with info on this trait.

tcharding reviewed (commented):
Man this is so thorough! I did not review the taproot example, many of the comments on the segwit one will apply I'm guessing. I'll give you time to use/ignore my suggestions then review the taproot one.

Top work bro!

realeinherjar commented:
> +1 to squashing into one commit -- though if you wanna be really nice, you could maybe split it into sections and make each one its own commit, or something.

I thought you guys did "squash merge" here. But I can squash it sure.

We don't squash merge. We try to follow:
- One patch per discreet change
- Rebase after review adding review changes to original patch
- Don't change code in a later patch that was changed in a previous one (so that reviewers don't comment on an early patch just to find its changed later on)

EDIT (tcharding): Ouch, I managed to edit your post yesterday instead of replying, sorry about that. FTR the answer here is written by me (tcharding). Sorry @realeinherjar, was not on purpose.

tcharding commented:
> +1 to squashing into one commit -- though if you wanna be really nice, you could maybe split it into sections and make each one its own commit, or something.

This PR is already bucket loads of work, speaking for myself I'd prefer you use your effort for real fixes to the cookbook rather than worrying about rebasing. This is example/docs code, its not as critical that we catch mistakes (which is a large part of what spitting it up helps achieve).

tcharding commented:
We have "do not automatically run CI for first time contributors" configured, if you want to use CI runs during dev of this you could throw up a trivial PR doing something else and we can merge it for you.

realeinherjar commented:
I see, I will fix something easy and then tag you as a reviewer.
But I am done for today, so tomorrow.

realeinherjar commented:
> I'll give you time to use/ignore my suggestions then review the taproot one.

I've added the segwit suggestions to the taproot one as well, since there was a lot of overlap.
It is ready to review.

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,338 @@
+    sighash_cache
+        .witness_mut(0)
+        .expect("valid witness")
+        .push_bitcoin_signature(&sig.serialize_der(), EcdsaSighashType::All);
```
pubkey should be pushed here after the signature.

realeinherjar replied:
fixed

tcharding commented:
> I've added the segwit suggestions to the taproot one as well, since there was a lot of overlap. It is ready to review.

Great, will review - likely won't get to it today but I'll get to it. Thanks man.

tcharding commented:
A quick rebase and force push and CI should run for you :)

realeinherjar commented:
@tcharding note that the new GitHub CI for GH Pages replaces the bash script you've made (I'm deleting it).

Additionally, I think you've done everything necessary to host a GH Page in a custom domain.
I added a CNAME file following the instructions here: https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/managing-a-custom-domain-for-your-github-pages-site#about-custom-domain-configuration

Could you check them to see if we need to do anything else.
I would be glad to help in whatever you need.

realeinherjar commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,338 @@
+    sighash_cache
+        .witness_mut(0)
+        .expect("valid witness")
+        .push_bitcoin_signature(&sig.serialize_der(), EcdsaSighashType::All);
```
```suggestion
        .push_bitcoin_signature(&sig.serialize_der(), EcdsaSighashType::All)
        .push(sk.public_key(&secp));
```

You mean this @tcharding?

tcharding replied:
Need to serialize the pubkey, see https://github.com/rust-bitcoin/rust-bitcoin/pull/2084/files

realeinherjar replied:
Solvet in f5d3fbda03ba7b07e21f8d256a9028872eaa6122

realeinherjar commented:
Updated the examples from https://github.com/rust-bitcoin/rust-bitcoin/pull/2095.
This is going to fail until we deal with the blockers in the PR description.

tcharding commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,341 @@
+    // Update the witness stack
+    let signature = Signature { sig, hash_ty: EcdsaSighashType::All };
+    let pk = sk.public_key(&secp);
+    *sighash_cache.witness_mut(input_index).unwrap() = Witness::p2wpkh(&signature, &pk);
```
Excuse me if I'm being slow but I'm confused as to why this in here, we run the tests against the latest release v.0.30.0 not against master?

realeinherjar replied:
Reverted back to `0.30`

tcharding commented:
I'm not getting what you are trying to achieve with the blockers? Are you thinking this won't merge until after the next release?

realeinherjar commented:
> I'm not getting what you are trying to achieve with the blockers? Are you thinking this won't merge until after the next release?

I can revert back.
I thought you were worried that the example didn't added the public key as you've said:

> Need to serialize the pubkey, see https://github.com/rust-bitcoin/rust-bitcoin/pull/2084/files

Since, this linked to the new `Witness::p2wpkh` constructor I thought we need to change the cookbook here.

But we can merge with the current `rust-bitcoin` version and update this later.

tcharding commented:
oh sorry mate if I caused confusion. I meant to just link you to code that worked to show what was needed :) To clarify so we are both on the same page, I think the cookbook should depend on the latest released version and we should use the soon-to-merge code in `examples` to have the same thing depending on master.

realeinherjar commented:
> oh sorry mate if I caused confusion. I meant to just link you to code that worked to show what was needed :) To clarify so we are both on the same page, I think the cookbook should depend on the latest released version and we should use the soon-to-merge code in `examples` to have the same thing depending on master.

Ok, gotcha.
I will revert back the commits.
In the mean time, how do we fix the public key serialization in `0.30`?

tcharding commented:
> In the mean time, how do we fix the public key serialization in 0.30?

The serialization is correct in 0.30 its just that the API is easy to mis-use. As an example the code `tx_taproot.rs` is correct because we use `TapSighashType::Default`, if a user inadvertantly copied this code then changed the sighash type the serialization would be wrong (missing the sighash type) - hence in https://github.com/rust-bitcoin/rust-bitcoin/pull/2095/files (and the new function in https://github.com/rust-bitcoin/rust-bitcoin/pull/2097/files#diff-ea912e7eb42db72b208cad947587783db67cfb40e77075e254420fbd774bf312) I used `to_vec` to get the correct behaviour irrespective of the sighash type - I think we should update this to do the same.

FTR one of the `rust-bitcoin` projects stated aims is to design APIs that are easy to use, hard to misuse, and don't require intimate knowledge of the bips or Bitcoin Core to use correctly. The `Witness::push` function is not currently conforming to these goals - now I write this I think we need to do more constructors and add docs to `push` saying "you should know _exactly_ what you are doing if you push things onto the witness with this function".

tcharding commented:
I knew this cookbook was a good idea and would shake things out.

tcharding commented:
Just noticed that there is a commit with a mention in the git brief message - `eb6c2eb fix: all @tcharding suggestions and fixes`.

(Is this some sort of auto-generated patch by github?)

Just to clarify, how we tend to do things is:
- each patch is a single conceptual change
- after review, if changes are needed, rebase, fix the original patch, and force push
- each patch needs a well formed description (describe why the change is needed then describe then change)

I hope my previous reviews did not lead you to the current git logs, I am trying to get better at reviewing :)

realeinherjar commented:
> FTR one of the `rust-bitcoin` projects stated aims is to design APIs that are easy to use, hard to misuse, and don't require intimate knowledge of the bips or Bitcoin Core to use correctly. The `Witness::push` function is not currently conforming to these goals - now I write this I think we need to do more constructors and add docs to `push` saying "you should know _exactly_ what you are doing if you push things onto the witness with this function".

Totally agree with this.



> Just noticed that there is a commit with a mention in the git brief message - `eb6c2eb fix: all @tcharding suggestions and fixes`.

I did it. But I've rebase everything to 2 commits :)


@tcharding this should be ready to review and merge now.

tcharding commented:
In this repo I'd like to maintain the same merge policy as in the rest of `rust-bitcoin` org, that means:
1. Only merge PRs using the bitcoin merge script: https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/main/github-merge.py
2. (1) implies we use a merge commit when merging which leads to (3)
3. Discreet, well described commits, that do a single thing.

For (3) we want the format of the git message to be "describe why the change is needed" then "describe what the change is" (if trivial the git brief message can act as the "what").

So, can you please update the commit logs of the 2 patches? (And please don't use mentions (@superstar) in the git logs because doing so causes GitHub to spam the user with notifications.

sanket1729 commented on cookbook/src/tx_segwit-v0.md:
````diff
@@ -0,0 +1,351 @@
+This is the `cargo` commands that you need to run this example:
+
+```bash
+cargo add bitcoin --features "std, rand-std"
````
@tcharding, I think this is bug in rust-bitcoin. I would expect `rand-std` to include `std`.

sanket1729 replied:
https://github.com/rust-bitcoin/rust-bitcoin/pull/2102

sanket1729 commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,351 @@
+This will be useful to mock a sender.
+In a real application these would be actual secrets.
+Under the hood we are using the [`secp256k1`](https://github.com/rust-bitcoin/rust-secp256k1/) crate to generate the key pair.
+`rust-secp256k1` is a wrapper around [libsecp256k1](https://github.com/bitcoin-core/secp256k1), a C
```
nit: Please don't bother changing unless there are other things that you are changing anyways.

This mention of secp and rust-secp seems like a good candidate for a footnote  and not directly relevant to the code.

realeinherjar replied:
Thanks for the suggestion. Done in f89936122927a1add2f309316ee924c644329c85

realeinherjar replied:
Both in Segwit and Taproot

sanket1729 commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,351 @@
+    // The input for the transaction we are constructing.
+    let input = TxIn {
+        previous_output: dummy_out_point,
+        script_sig: script_code.clone(),
```
script_sig should be empty. (ScriptBuf::new() ) `script_code` is only used for signing.

realeinherjar replied:
Done in f89936122927a1add2f309316ee924c644329c85
Not applicable to Taproot since it already uses `Scriptbuf::new`

sanket1729 commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,351 @@
+        .push(&pk.serialize());
+
+    // Print the transaction ready to broadcast
+    let tx = sighash_cache.into_transaction();
```
While `sighash_cache::witness_mut` this is a fine way to do sighash operations, I feel that it really non-intuitive. It is primarily to avoid allocations when iterating through tx inputs while creating witnesses. 

I feel the below is better for a tutorial/cookbook.

```rust
    // Update the witness wpkh stack: signature + public key.
    let pk = sk.public_key(&secp);
    let mut witness = &mut tx.input[0].witness;

    witness.push_bitcoin_signature(
        &sig.serialize_der(),
        EcdsaSighashType::All
    );
    witness.push(&pk.serialize());
```

realeinherjar replied:
I've updated how we are mutating the witness in both segwit and taproot with your suggestion.
Can you check it now please?

sanket1729 commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,351 @@
+Inside the [`TxIn`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.TxIn.html) struct we are setting the following fields:
+
+- `previous_output` is the outpoint of the dummy UTXO we are spending; it is a [`OutPoint`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.OutPoint.html) type.
+- `script_sig` is the script code required to spend a P2WPKH output; it is a [`ScriptBuf`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/script/struct.ScriptBuf.html) type.
```
As mentioned above, this needs to be modified.

realeinherjar replied:
Done in f89936122927a1add2f309316ee924c644329c85

sanket1729 commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,351 @@
+    // The transaction we want to sign and broadcast.
+    let unsigned_tx = Transaction {
+        version: 2,
+        lock_time: absolute::LockTime::from_height(0).expect("valid height"),
```
We should do `absolute::LockTime::ZERO`

realeinherjar replied:
Good catch, makes it much more simpler.
Done in f89936122927a1add2f309316ee924c644329c85,
also updated the explanation (both in Segwit and Taproot).

sanket1729 commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,351 @@
+   We are using the `const DUMMY_UTXO_AMOUNT` that we defined earlier.
+- `sighash_type` is the type of sighash; it is a [`EcdsaSighashType`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html) enum.
+   We are using the [`All`](https://docs.rs/bitcoin/latest/bitcoin/sighash/enum.EcdsaSighashType.html#variant.All) variant,
+   which indicates that the sighash will include all the outputs.
```
all inputs and outputs

realeinherjar replied:
Doone in f89936122927a1add2f309316ee924c644329c85 (both sw and tr)

sanket1729 commented on cookbook/src/tx_segwit-v0.md:
```diff
@@ -0,0 +1,351 @@
+
+In `let change = TxOut {...}` we are instantiating the change output.
+It is very similar to the `spend` output, but we are now using the `const CHANGE_AMOUNT` that we defined earlier.
+And also we are locking the output to an address that we control:
```
nit: Can we also add a footnote here? We are sending to the same change address that we received funds from. ( i.e same wpkh is used in spend as well as in change) 

In practice, the change address would be a different address.

realeinherjar replied:
Footnote added to both sw and tr in f89936122927a1add2f309316ee924c644329c85.

sanket1729 reviewed (commented):
Only reviewed segwit example. I think most comments are nit and can be addressed later if needed, but the script_code one should be addressed.

realeinherjar commented:
@sanket1729 thanks for the review and suggestions.
I agreed with everything and incorporate them all.
Can you please check the witness updating procedure now in both segwit and taproot.

Here are the permalinks for easy visualization

Segwit:

https://github.com/rust-bitcoin/www.rust-bitcoin.org/blob/f15e6806a1bccdf9cf817371c858c67508632532/cookbook/src/tx_segwit-v0.md?plain=1#L208-L229

Taproot
https://github.com/rust-bitcoin/www.rust-bitcoin.org/blob/f15e6806a1bccdf9cf817371c858c67508632532/cookbook/src/tx_taproot.md?plain=1#L212-L241

tcharding reviewed (approved):
ACK 2993048453dd94852863a2cf363604449784cb41

tcharding commented:
I've acked but can you re-write the commit logs before this is considered for merge please. Also I'm not totally confident in my ability to review the github actions stuff, especially the deployment, but if it works its way better than the script I had hacked up. We will find out soon enough when we merge, YOLO!

tcharding commented:
FTR I served the cookbook locally and it looks awesome!

sanket1729 reviewed (approved):
ACK 2993048453dd94852863a2cf363604449784cb41. Just reviewed the examples for correctness and verified that I can serve the files locally. 

I am not too familiar with github actions stuff going on here, but this is good to go from my side.

realeinherjar commented:
Thanks for the reviews and for dedicating time to help in the making of this content.

I am 99% certain on the GitHub Pages deployment.
However, once we fix this we don't need to touch it or run shell scripts again to build a new version of the site.
If anything goes wrong with the site I'll fix it straight away.

@tcharding I've squashed all to a single commit and changed the commit message.
See if you want me to change anything else.

tcharding reviewed (approved):
ACK cee495382e34859db4e7db0f800a113a07901bf0

tcharding commented:
Looks good, lets go!

tcharding commented:
I checked the diff since you reviewed @sanket1729 and its only changes to the github actions etc., no code changes to the examlpes.

I'll let this sit over the weekend then merge it unless anyone wants more time to look.

sanket1729 commented:
utACK [cee4953](https://github.com/rust-bitcoin/www.rust-bitcoin.org/commit/cee495382e34859db4e7db0f800a113a07901bf0). Go ahead

tcharding commented:
Doesn't look like this has been deployed. I cleared my cache and checked the site. Did not investigate further (its the weekend here).

realeinherjar commented:
Ok I was asleep. Looks like some Hugo error. I will debug it and submit a fix in the next hours.
That came out of the blue.

realeinherjar commented:
Fix in rust-bitcoin/www.rust-bitcoin.org#14 @tcharding 
(Sorry about that)

tcharding commented:
No sweat, thanks for the fix. Was cool to wake up and have this fixed already.

realeinherjar commented:
@tcharding the website builds ok now.
But we need someone with maintainer access (needs to be able to set the repo settings) to activate GitHub pages.

The instructions are here: https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-from-a-branch
We need to "Publish from a branch" and the defaults `gh-pages` and `/ (root)` will suffice.

Finally, since it is a custom domain you might need to do some tweaks explained here: https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/about-custom-domains-and-github-pages

EDIT: If you are struggling with this, let me know. We can jump in a quick call and I can guide you through if necessary...

tcharding commented:
Oh, this is my fault for not remembering how I set this up. How I had the build script and the repos set up doesn't work with the actions stuff we just merged. I had it so that this repo was just the source code and the actual site, once built, was pushed to https://github.com/rust-bitcoin/rust-bitcoin.github.io/tree/master. That repo has the custom domain etc. configured already. The reasoning behind this split was so that the source code was not entangled with the hosting service (currently github pages) so that if we wanted to move to a different host it was trivial to do. Other maintainers, if I remember correctly, didn't really see the need for this so I guess we could just delete the other repo and configure github pages from this repo. I'm fasting at the moment and in general I really hate doing any sort of web development, but I'll come back to this once I'me eating again.

realeinherjar commented:
As always if you need an extra pair of hands let me know...
I agree that `https://github.com/rust-bitcoin/rust-bitcoin.github.io` should be deleted and replaced by `https://github.com/rust-bitcoin/www.rust-bitcoin.org`.
